feat: add parallel constraint validation (experimental)

david-waltermire opened the pull request:
## Summary

Adds experimental parallel constraint validation with a `--threads N` CLI option. When enabled, sibling nodes are validated concurrently using a ForkJoinPool, potentially improving validation performance on large documents with multi-core systems.

## Changes

### New Files
- `ParallelValidationConfig.java` - Thread pool management with lazy initialization and proper lifecycle
- `ParallelValidationConfigTest.java` - Unit tests for configuration
- `DefaultConstraintValidatorThreadSafetyTest.java` - Concurrency tests
- `FindingCollectingConstraintValidationHandlerTest.java` - Thread-safety tests
- `DynamicContextTest.java` - Execution stack isolation tests
- `PRDs/20251217-parallel-validation/` - Design document and implementation plan

### Modified Files
- `DynamicContext.java` - Thread-safe with per-context execution stacks
- `DefaultConstraintValidator.java` - Parallel sibling traversal
- `FindingCollectingConstraintValidationHandler.java` - Thread-safe findings collection
- `ValidationFeature.java` - Added `PARALLEL_THREADS` feature
- `IBindingContext.java` - Read parallel config from validation features
- `AbstractValidateContentCommand.java` - Added `--threads` CLI option

## Key Design Decisions

1. **ForkJoinPool**: Uses work-stealing to handle nested parallelism without deadlock
2. **PARALLEL_THRESHOLD = 4**: Only parallelize when >= 4 siblings (fixed, not configurable)
3. **Deterministic output**: Findings sorted by Metapath for consistent results
4. **Graceful degradation**: Falls back to sequential on interruption
5. **Experimental warning**: Prints warning to stderr when `--threads > 1`

## Thread Safety Approach

- `ConcurrentHashMap` for shared state (`valueMap`, `availableDocuments`, `letVariableMap`)
- `ConcurrentLinkedQueue` for findings collection
- `AtomicReference` for highest severity tracking
- Per-context execution stacks (copied from parent on `subContext()`)
- Synchronized lists for index key references
- Future cancellation on exception to prevent resource leaks

## Usage

```bash
# Sequential validation (default, current behavior)
metaschema validate document.json

# Parallel validation with 4 threads (experimental)
metaschema validate --threads 4 document.json
```

## Test Plan

- [x] Unit tests for ParallelValidationConfig
- [x] Thread-safety tests for DynamicContext
- [x] Concurrency tests for FindingCollectingConstraintValidationHandler
- [x] Behavioral equivalence tests (parallel produces same results as sequential)
- [x] Index accumulation thread-safety tests
- [x] Allowed values tracking thread-safety tests
- [x] Input validation for --threads option (NumberFormatException, value >= 1)
- [x] All 4326 existing tests pass
- [x] Manual testing with NIST CSF catalog (verified same results with/without threads)

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Experimental opt-in parallel constraint validation via a new --threads CLI option; sequential mode remains default and a warning is shown when parallelism is used.

* **Improvements**
  * Thread-safety and concurrency hardening across validation for reliable parallel runs, deterministic findings ordering, and robust cancellation/exception propagation.

* **Tests**
  * Added concurrency and deterministic-ordering tests to validate parallel and sequential behavior.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

david-waltermire committed:
docs: add PRD for parallel constraint validation

Add design document and implementation plan for experimental parallel
constraint validation feature with --threads CLI option.

Key features:
- ParallelValidationConfig for thread pool management
- Thread-safe shared state in validators
- Parallel sibling traversal in constraint validator
- --threads N CLI option with experimental warning

david-waltermire committed:
docs: address PR review comments

- Add @throws IllegalArgumentException to withThreads() Javadoc
- Add future cancellation on exception in parallel traversal to avoid
  resource leaks when validation fails
- Fix grammar: "CLI interface" → "CLI"
- Add language specifiers to code blocks

david-waltermire committed:
feat: add experimental parallel constraint validation

Add --threads CLI option for parallel constraint validation with
configurable thread count. This is an experimental feature that
enables concurrent validation of assembly children.

Key changes:
- Add ParallelValidationConfig for thread pool management
- Make DynamicContext thread-safe with subContext() method
- Make FindingCollectingConstraintValidationHandler thread-safe
- Make DefaultConstraintValidator thread-safe with parallel traversal
- Add PARALLEL_THREADS ValidationFeature
- Add --threads option to validate-content CLI command

When --threads > 1, a warning is displayed indicating the feature
is experimental. Default behavior (--threads 1) uses sequential
validation for backward compatibility.

david-waltermire committed:
fix: use ForkJoinPool to prevent deadlock with nested parallelism

Fixed thread pools deadlock when all threads wait for their children
to complete via future.get(). ForkJoinPool handles nested parallelism
via work-stealing, preventing deadlock with deeply nested structures.

aj-stein reviewed (dismissed):
Nice work, could come in handy with bigger document collections.

david-waltermire committed:
Add input validation for --threads CLI option

- Handle NumberFormatException for non-integer values
- Validate thread count is at least 1
- Provide clear error messages with INVALID_ARGUMENTS exit code

david-waltermire committed:
docs: add withThreads() validation implementation to PRD

david-waltermire commented on PRDs/20251217-parallel-constraint-validation/PRD.md:
```diff
@@ -0,0 +1,481 @@
+      for (Future<?> future : futures) {
+        future.get();
+      }
+    } catch (ExecutionException e) {
+      cancelRemainingFutures(futures);
+      Throwable cause = e.getCause();
+      if (cause instanceof RuntimeException) {
+        throw (RuntimeException) cause;
+      }
+      throw new ConstraintValidationException(cause);
+    } catch (InterruptedException e) {
+      cancelRemainingFutures(futures);
+      Thread.currentThread().interrupt();
+      throw new ConstraintValidationException("Validation interrupted", e);
+    }
```
Agreed.

david-waltermire replied:
fixed

david-waltermire commented on PRDs/20251217-parallel-constraint-validation/PRD.md:
```diff
@@ -0,0 +1,481 @@
+  /**
+   * Get the executor, creating an internal pool if needed.
+   *
+   * @return the executor service
+   */
+  @NonNull
+  ExecutorService getOrCreateExecutor();
+
+  /**
+   * Shut down internal executor if one was created.
+   * Does nothing if using an external executor.
+   */
+  void shutdownInternalExecutor();
+}
```
Will update the PRD

david-waltermire committed:
refactor: use ConstraintValidationException for parallel validation errors

- Use ExceptionUtils.wrap() pattern for checked exception handling
- Align PRD method names with implementation (getExecutor, close)
- Use AutoCloseable/try-with-resources pattern in CLI example
